agent/src/com/cloud/agent/direct/download/DirectTemplateDownloaderImpl.java

```diff
@@ -149,36 +154,16 @@ public DirectTemplateInformation getTemplateInformation() {
         return new DirectTemplateInformation(installPath, size, checksum);
     }
 
-    /**
-     * Return checksum command from algorithm
-     */
-    private String getChecksumCommandFromAlgorithm(String algorithm) {
-        if (algorithm.equalsIgnoreCase("MD5")) {
-            return "md5sum";
-        } else if (algorithm.equalsIgnoreCase("SHA-1")) {
-            return "sha1sum";
-        } else if (algorithm.equalsIgnoreCase("SHA-224")) {
-            return "sha224sum";
-        } else if (algorithm.equalsIgnoreCase("SHA-256")) {
-            return "sha256sum";
-        } else if (algorithm.equalsIgnoreCase("SHA-384")) {
-            return "sha384sum";
-        } else if (algorithm.equalsIgnoreCase("SHA-512")) {
-            return "sha512sum";
-        } else {
-            throw new CloudRuntimeException("Unknown checksum algorithm: " + algorithm);
-        }
-    }
-
     @Override
     public boolean validateChecksum() {
         if (StringUtils.isNotBlank(checksum)) {
-            ChecksumValue providedChecksum = new ChecksumValue(checksum);
-            String algorithm = providedChecksum.getAlgorithm();
-            String checksumCommand = "echo '%s %s' | %s -c --quiet";
-            String cmd = String.format(checksumCommand, providedChecksum.getChecksum(), downloadedFilePath, getChecksumCommandFromAlgorithm(algorithm));
-            int result = Script.runSimpleBashScriptForExitValue(cmd);
-            return result == 0;
+            try {
+                return DigestHelper.check(checksum, new FileInputStream(downloadedFilePath));
+            } catch (IOException e) {
+                throw new CloudRuntimeException("could not check sum for file: " + downloadedFilePath);
```

Comment:
missing second argument to view the actual exception message/stack trace:
`file: " + downloadedFilePath, e)`

Reply:
yes, I agree. though people have been pushing towards not including root causes in exception traces i will add it!

Reply:
come to think of it it does not make as much sense in the NoSuchAlgorithmException case. Doe you agree @marcaurele ?

Reply:
IMO if the exception is due to some system/programming problem, the stack should be given to give a clear hint to people where to look for the error in the code. If the exception is aimed at the end user (API end user), then it does not make sens as it is a purely parameter, state,... error. So i'll keep the root cause for the NoSuchAlgorithmException.

Reply:
ok, I thought the error is purely the chosen algorithnm so don't include the root cause.